File: Problem/By section/Greedy/Problem 665/problem_665.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // Time: O(n), Space: O(1)
    bool checkPossibility(vector<int>& nums) {
        int size = nums.size();
        if(size < 2) return true;

        int count = 0;
        for(int i = 1; i < size && count < 2; i++) {
            if(nums[i] < nums[i - 1]) {
                count++;

                // for a sub sequence [a1, a2, a3, a4]
                // we have a2 > a3
                // 1. if a1 < a3, then we can change a2 to a3, result array [a1, a3, a3, a4] will be valid
                // 2. if a2 < a4, then we can change a3 to a2, result array [a1, a2, a2, a4] will be valid
                // 3. Otherwise, we can not resolve the conflict
                if(count > 1 ||
                    (
                        (i > 1 && nums[i - 2] > nums[i]) &&
                        (i < size - 1 && nums[i - 1] > nums[i + 1])
                    )
                ) {
                    return false;
                }
            }
        }

        return true;
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> input = {5, 7, 1, 8};
    cout << s.checkPossibility(input) << endl;
    return 0;
}

File: Problem/By section/Greedy/Problem 665/problem_665.java
class Solution {
    // Time: O(n), Space: O(1)
    public boolean checkPossibility(int[] nums) {
        if(nums.length == 0) return true;
        
        // count for the change of number
        int count = 0;
        
        for(int i = 1; i < nums.length && count < 2; i++){
            // if it is non-decreasing
            if(nums[i] >= nums[i - 1]) continue;

            count++;

            // we defaultly may change the nums[i - 1] to nums[i]
            // since it wont demage the non-decreasing order of [i,i + 1,.....]
            // while if nums[i - 2] is greater than nums[i]
            // we have change nums[i] to nums[i - 1] in order to match the non-decreasing order of [0, 1, ...., i - 2, i - 1]
            if(i - 2 >= 0 && nums[i - 2] > nums[i]){
                nums[i] = nums[i - 1];
            }else{
                nums[i - 1] = nums[i];
            }
        }
        
        return count < 2;
    }
}

public class problem_665{
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.checkPossibility(new int[]{3,4,2,3}));
    }
}